### ISSUE TITLE
CUDA out of memory, No Checkpoint or Snapshot Saved After Press "Save weight" Button during Training or Finish Training!

### ISSUE THREAD

**Issue by Javadst** [d8ahazard/sd_dreambooth_extension#955]
### Kindly read the entire form below and fill it out with the requested information.

**Please find the following lines in the console and paste them below. _If you do not provide this information, your
issue will be automatically closed._**

`

Python revision: 3.10.6 &
Python revision: 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Dreambooth revision: 43ae9d55531004f1dedaea7ac2443e9b16739913
SD-WebUI revision:

Checking Dreambooth requirements...
[+] bitsandbytes version 0.35.0 installed.
[+] diffusers version 0.10.2 installed.
[+] transformers version 4.25.1 installed.
[+] xformers version 0.0.14.dev0 installed.
[+] torch version 1.13.1+cu117 installed.
[+] torchvision version 0.14.1+cu117 installed.

**Have you read the [Readme](https://github.com/d8ahazard/sd_dreambooth_extension#readme)?**
Yes
**Have you completely restarted the stable-diffusion-webUI, not just reloaded the UI?**
Yes
**Have you updated Dreambooth to the latest revision?**
Yes
**Have you updated the Stable-Diffusion-WebUI to the latest version?**
Yes
**No, really. Please save us both some trouble and update the SD-WebUI and Extension and restart before posting this.
Reply 'OK' Below to acknowledge that you did this.**
OK
**Describe the bug**

After Press "Save weight" Button or after the end of training i have this error and No Checkpoint or Snapshot Saved.
If  "UNET: True" I get this error message as soon as finished create the class images and start training : (CUDA out of memory).

**Provide logs**

If a crash has occurred, please provide the *entire* stack trace from the log, including the last few log messages *
before* the crash occurred.

**Environment**

What OS?
10
If Windows - WSL or native?
native
What GPU are you using?
Asus 3080 10GB VRam

**Screenshots/Config**
  ***** Running training *****
  Num batches each epoch = 80
  Num Epochs = 150
  Batch Size Per Device = 1
  Gradient Accumulation steps = 1
  Total train batch size (w. parallel, distributed & accumulation) = 1
  Text Encoder Epochs: 150
  Total optimization steps = 6000
  Total training steps = 12000
  Resuming from checkpoint: False
  First resume epoch: 0
  First resume step: 0
  Lora: False, Adam: True, Prec: fp16
  Gradient Checkpointing: True
  EMA: True
  UNET: False
  Freeze CLIP Normalization Layers: False
  LR: 2e-06
  V2: False
Steps:  65%|██████████████████████████████████▎                  | 7760/12000 [42:30<23:13,  3.04it/s, loss=0.00208, loss_avg=0.106, lr=2e-6, vram_usage=6.1]Removing log: G:\AI\Application\stable-diffusion-webui-master\models\dreambooth\_x_Fati_x_\logging\db_log_2023-02-17-22-08-47.log
OOM Detected, reducing batch/grad size to 0/1.
Traceback (most recent call last):
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\memory.py", line 121, in decorator
    return function(batch_size, grad_size, prof, log_file, *args, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 951, in inner_loop
    check_save(progress_bar, True)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 589, in check_save
    save_weights(save_image, save_model, save_snapshot, save_checkpoint, save_lora, pbar)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 639, in save_weights
    s_pipeline = s_pipeline.to(accelerator.device)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\diffusers\pipeline_utils.py", line 270, in to
    module.to(torch_device)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\transformers\modeling_utils.py", line 1682, in to
    return super().to(*args, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 989, in to
    return self._apply(convert)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 641, in _apply
    module._apply(fn)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 641, in _apply
    module._apply(fn)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 641, in _apply
    module._apply(fn)
  [Previous line repeated 3 more times]
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 664, in _apply
    param_applied = fn(param)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 987, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 10.00 GiB total capacity; 8.65 GiB already allocated; 0 bytes free; 8.79 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Steps:  65%|██████████████████████████████████▎                  | 7760/12000 [43:10<23:35,  3.00it/s, loss=0.00208, loss_avg=0.106, lr=2e-6, vram_usage=6.1]
Traceback (most recent call last):
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\scripts\dreambooth.py", line 455, in start_training
    result = main(config, use_txt2img=use_txt2img)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 989, in main
    return inner_loop()
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\memory.py", line 119, in decorator
    raise RuntimeError("No executable batch size found, reached zero.")
RuntimeError: No executable batch size found, reached zero.
Restored system models.

**Comment by Arilziem** [comment]
I am pretty sure the 'Save weights' button will simply save the current state of the dreambooth model from VRAM to disk at the end of the epoch (although that could probably be made clearer in the UI). It has nothing to do with checkpoints or (LORA) snapshots, that is a different step.

Your OOM is probably because 10 GB VRAM is not enough to train UNET and text encoder together (unless you use LORA). At least I never managed to on my 3080. Set the text encoder slider below 'train unet' to 0 and it should work. You will need to disable EMA as well for non-LORA.

Also, memory management is somewhat broken in main branch at the moment, so until dev gets merged back into main (probably very soon though from the looks of it), you will probably have to disable sample images as well.

**Comment by Javadst** [comment]
Thanks for the help, but as soon as I hit this button at 65% it stopped and nothing was saved and then it gave this error message.I tried this several times.
I disabled Unet, (if it's on it doesn't get to the training stage at all.)
OK, I'll disable the EMA and test again.
Question: Will disabling Unet, EMA make a significant difference in output quality?

**Comment by Arilziem** [comment]
Normally, you do not need to press that button. I have used it once when I scheduled 20 epochs of training (and no saving in between), but wanted to copy the model after 10 epochs as a backup.

The UNET / text encoder settings are a bit confusing. When you disable UNET, you will only train the text encoder, which in most cases is not enough. When you enable UNET training, by default you will have the slider right below it set to something around 0.2, which means it will train the UNET AND the text encoder at a ratio of 1:0.2, so if you have e.g. 10 epochs of total training, it will train the UNET for 10 and the text encoder for 2. That will not work with 10 GB VRAM, so you need to set that to 0.

If you do not train UNET at all, your model will not learn new imagery. From my (limited) understanding, the text encoder will basically learn to associate what the model thinks is in your training images ("latents") with the captions you provide. You want UNET almost always.

Training the text encoder is optional, but can really help get better results. I usually train UNET for 10-20 epochs, then text encoder for 1-2 epochs (by unchecking UNET), then again UNET ... until I like the results.

EMA can improve quality a bit from what I know, but I have gotten good results without it.

**Comment by Javadst** [comment]
Thank you for your time and detailed explanation.
I disabled EMA and UNET, then tested the "Save Weight" button, it worked Great, then after the 22 epoch was over, I canceled the training and the checkpoint saved fine. :)
After that, I enabled UNET and set the text encoder to 0, Resume Training gives me this error:

  ***** Running training *****
  Num batches each epoch = 80
  Num Epochs = 150
  Batch Size Per Device = 1
  Gradient Accumulation steps = 1
  Total train batch size (w. parallel, distributed & accumulation) = 1
  Text Encoder Epochs: 0
  Total optimization steps = 6000
  Total training steps = 12000
  Resuming from checkpoint: True
  First resume epoch: 22
  First resume step: 1708
  Lora: None, Adam: True, Prec: fp16
  Gradient Checkpointing: True
  EMA: False
  UNET: True
  Freeze CLIP Normalization Layers: False
  LR: 2e-06
  V2: False
Steps:   0%|                      | 4/10292 [00:21<9:01:56,  3.16s/it, loss=nan, loss_avg=nan, lr=1e-6, vram_usage=8.7]Removing log: G:\AI\Application\stable-diffusion-webui-master\models\dreambooth\_x_Fati_x_\logging\db_log_2023-02-18-02-05-35.log
OOM Detected, reducing batch/grad size to 0/1.
Traceback (most recent call last):
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\memory.py", line 121, in decorator
    return function(batch_size, grad_size, prof, log_file, *args, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 877, in inner_loop
    accelerator.backward(loss)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\accelerate\accelerator.py", line 1314, in backward
    self.scaler.scale(loss).backward(**kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\autograd\__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\autograd\function.py", line 267, in apply
    return user_fn(self, *args)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\utils\checkpoint.py", line 157, in backward
    torch.autograd.backward(outputs_with_grad, args_with_grad)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\autograd\__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 52.00 MiB (GPU 0; 10.00 GiB total capacity; 8.56 GiB already allocated; 0 bytes free; 8.77 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Steps:   0%|                     | 4/10292 [00:23<16:43:33,  5.85s/it, loss=nan, loss_avg=nan, lr=1e-6, vram_usage=8.7]
Traceback (most recent call last):
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\scripts\dreambooth.py", line 455, in start_training
    result = main(config, use_txt2img=use_txt2img)
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\train_dreambooth.py", line 989, in main
    return inner_loop()
  File "G:\AI\Application\stable-diffusion-webui-master\extensions\sd_dreambooth_extension\dreambooth\memory.py", line 119, in decorator
    raise RuntimeError("No executable batch size found, reached zero.")
RuntimeError: No executable batch size found, reached zero.
Restored system models.

**Comment by Arilziem** [comment]
No problem, glad I could help.

Since you still experience OOM with UNET training, make sure you have attention set to 'xformers' (and that you have either version 0.14 or 0.17 installed, not 0.16), or 'flash_attention'. If not, you might have to start with a fresh model. If you already have attention configured correctly, you might have too many other programs running.

If you need further assistance with the right settings, I would suggest you close this issue and post in the discussions section, since this seems to be a configuration issue rather than a bug. You should also post your training configuration (see [bug reporting template](https://github.com/d8ahazard/sd_dreambooth_extension/blob/main/.github/ISSUE_TEMPLATE/bug_report.md), 'Screenshots/Config'), that makes it easier to spot problems.

Good luck!

**Comment by Javadst** [comment]
Thank you.
I was using firefox and according to your instructions I used chrome and the problem was solved.
But after finishing training, it presented a black image and after restarting the program, I encountered this error:
---------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "G:\AI\Application\stable-diffusion-webui-master\modules\call_queue.py", line 56, in f
    res = list(func(*args, **kwargs))
  File "G:\AI\Application\stable-diffusion-webui-master\modules\call_queue.py", line 37, in f
    res = func(*args, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\modules\txt2img.py", line 56, in txt2img
    processed = process_images(p)
  File "G:\AI\Application\stable-diffusion-webui-master\modules\processing.py", line 486, in process_images
    res = process_images_inner(p)
  File "G:\AI\Application\stable-diffusion-webui-master\modules\processing.py", line 628, in process_images_inner
    samples_ddim = p.sample(conditioning=c, unconditional_conditioning=uc, seeds=seeds, subseeds=subseeds, subseed_strength=p.subseed_strength, prompts=prompts)
  File "G:\AI\Application\stable-diffusion-webui-master\modules\processing.py", line 828, in sample
    samples = self.sampler.sample(self, x, conditioning, unconditional_conditioning, image_conditioning=self.txt2img_image_conditioning(x))
  File "G:\AI\Application\stable-diffusion-webui-master\modules\sd_samplers_kdiffusion.py", line 323, in sample
    samples = self.launch_sampling(steps, lambda: self.func(self.model_wrap_cfg, x, extra_args={
  File "G:\AI\Application\stable-diffusion-webui-master\modules\sd_samplers_kdiffusion.py", line 221, in launch_sampling
    return func()
  File "G:\AI\Application\stable-diffusion-webui-master\modules\sd_samplers_kdiffusion.py", line 323, in <lambda>
    samples = self.launch_sampling(steps, lambda: self.func(self.model_wrap_cfg, x, extra_args={
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\repositories\k-diffusion\k_diffusion\sampling.py", line 145, in sample_euler_ancestral
    denoised = model(x, sigmas[i] * s_in, **extra_args)
  File "G:\AI\Application\stable-diffusion-webui-master\venv\lib\site-packages\torch\nn\modules\module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "G:\AI\Application\stable-diffusion-webui-master\modules\sd_samplers_kdiffusion.py", line 139, in forward
    devices.test_for_nans(x_out, "unet")
  File "G:\AI\Application\stable-diffusion-webui-master\modules\devices.py", line 152, in test_for_nans
    raise NansException(message)
modules.devices.NansException: A tensor with all NaNs was produced in Unet. Use --disable-nan-check commandline argument to disable this check.
---------------------------------------------------------------------------
I started a new training once again and the same problem persists.
I didn't catch anything by searching for the error, because such a problem was known from xformers version 0.0.16, while my xformers version is 0.14.

**Comment by Arilziem** [comment]
Black image most likely means broken model, yes. The model was likely damaged during the OOM issues. Make sure you create a new model before starting training. If the problem persists, please open a new thread in Discussion or a new issue.

Since the original issues seem to be resolved, it would be great if you could close this issue as resolved as well. Thanks!

**Comment by Javadst** [comment]
OK, Thank you very much.
